Client/game_sa/CFxSA.cpp

```diff
@@ -239,3 +246,23 @@ void CFxSA::TriggerFootSplash(CVector& vecPosition)
         call    dwFunc
     }
 }
+
+bool CFxSA::IsShadowsLimitReached() const
+{
+    unsigned short& CShadows_ShadowsStoredToBeRendered = *(unsigned short*)0xC403DC;
+    return CShadows_ShadowsStoredToBeRendered >= 48;
+}
+
+bool CFxSA::AddShadow(eShadowTextureType shadowTextureType, const CVector& vecPosition, const CVector2D& vecOffset1, const CVector2D& vecOffset2, SColor color, eShadowType shadowType, float fZDistance,
+                      bool bDrawOnWater, bool bDrawOnBuildings)
+{
+    if (IsShadowsLimitReached())
+        return false;
+
+    void*      textureAddress = *(void**)(SHADOW_BASE_TEXTURE_OFFSET + (int)shadowTextureType * 4);
```

Comment:
I wonder what will happen if you try `eShadowTextureType::PLANE = -1`. Also, you could probably just use pointer arithmetic on `RwTexture**` and avoid this funky pointer madness.